README
DBIx-DBH version 0.01
=====================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2004 by Terrence Brannon

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.




ChangeLog

	* Add SQLite driver (Mark Stosberg)

	* Allow new drivers to published and used without modifying DBIx::DBH (Mark Stosberg)

2004-12-01  Terrence Brannon  <terry@hcoop.net>

	* changed "vendor" key to driver per Cavaletto

	* added docs for form_dsn call.

Revision history for Perl extension DBIx::DBH.

0.01  Mon Nov 22 23:22:31 2004
	- original version; created by h2xs 1.23 with options
		-b 5.6.1 -A -X -n DBIx::DBH



lib/DBIx/DBH.pm
package DBIx::DBH;

use Moose;
use Moose::Util::TypeConstraints;
use DBI;

has [ 'username', 'password' ] => (is => 'rw', isa => 'Str');

subtype 'DSNHashRef'  => as 'HashRef'  => where { defined($_->{driver}) };

has 'dsn'  => (is => 'rw', isa => 'DSNHashRef', required => 1);
has 'attr' => (is => 'rw', isa => 'HashRef');


sub dsn_string {
  my($self)=@_;

  my %dsn = % { $self->dsn } ;
  my $driver = delete($dsn{driver});
  my $dsn = "dbi:$driver";

  $dsn .= ";$_=$dsn{$_}"  for ( sort keys %dsn );

  $dsn;
}

sub for_dbi {
  my($self)=@_;
  ($self->dsn_string, $self->username, $self->password, $self->attr);
}

sub for_rose_db {
  my($self)=@_;

  (
   username => $self->username,
   password => $self->password,
   %{$self->dsn}
  )
}

sub dbh {
  my($self)=@_;

  use DBI;

  my $dbh = DBI->connect($self->for_dbi)
}

sub conn {
  my($self)=@_;

  require DBIx::Connector;

  my $dbh = DBIx::Connector->new($self->for_dbi);

}


our $VERSION = '0.3';





1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

 DBIx::DBH - helper for DBI connection( data)?

=head1 SYNOPSIS

 use DBIx::DBH;

 my $config = DBIx::DBH->new
   (
     user => $user,
     pass => $pass,
     dsn  => { driver => 'mysql', port => 3306 },
     attr => { RaiseError => 1 }
   );

 $config->for_rose_db; # outputs data structure for Rose::DB::register_db
 $config->for_dbi;     # outputs data structure for DBI connect()
 $config->dbh;  # makes a database connection with DBI
 $config->conn; # makes a DBIx::Connector instance

=head1 ABSTRACT

L<DBIx::DBH> allows you to specify the DBI dsn ( L<DBI/"connect> )
as a hash ref instead of a string. A hashref is a more viable structure
in a few cases:

=over 4

=item * working with Rose::DB

L<Rose::DB::Tutorial/Registering_data_sources> shows that L<Rose::DB>
expects the dsn information as discrete key-value pairs as opposed to
a string. The C<< ->for_rose_db >> method takes the DBIx::DBH instance
and returns a hash array which can be consumed by L<Rose::DB/register_db>

=item * programmatic connection attempts

It is much easier to manipulate a hash programmatically if you need to 
systematically modify it as part of a series of connection attempts.

=item * high-level structure

Whether you are talking about configuration file utilities or form data,
most data from these modules comes back directly as hashes. So you have
a more direct way of shuttling data into a database connection if you 
use this module:

   my $dbh = DBIx::DBH->(dsn => $cgi->form_data->{dsn})->dbh;

Instead of a bunch of string twiddling.

=back

=head1 METHODS

=head1 Legacy Version

A procedural version of DBIx::DBH is still available as
L<DBIx::DBH::Legacy>.

=head1 SEE ALSO

=over

=item * L<Config::DBI>

=item * L<DBIx::Connect>

=item * L<DBIx::Password>

=item * L<Ima::DBI>

=back

=head2 Links

=head3 "Avoiding compound data in software and system design"

L<http://perlmonks.org/?node_id=835894>



=head1 AUTHOR

Terrence Brannon, E<lt>bauhaus@metaperl.comE<gt>

Sybase support contributed by Rachel Richard.

Mark Stosberg did all of the following:

=over

=item * contributed Sqlite support

=item * fixed a documentation bug

=item * made DBIx::DBH more scaleable

Says Mark: "Just as DBI needs no modifications for a new driver to work,
neither should this module.

I've attached a patch which refactors the code to address this.

Rather than relying on a hardcoded list, it tries to 'require' the
driver, or dies with a related error message.

This could lower your own maintenance effort, as others can publish
additional drivers directly without requiring a new release of
DBIx::DBH for it to work."

L<http://rt.cpan.org/Ticket/Display.html?id=18026>

=back



Substantial suggestions by M. Simon Ryan Cavaletto.

=head1 SOURCECODE

L<http://github.com/metaperl/dbix-dbh>

=head1 COPYRIGHT AND LICENSE

Copyright (C) by Terrence Brannon

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/DBIx/DBH/mysql.pm
package DBIx::DBH::mysql ;

use base qw(DBIx::DBH);
use Params::Validate qw( :all );

@optional_mysql_ = qw(mysql_client_found_rows
		      mysql_compression
		      mysql_connect_timeout
		      mysql_read_default_file
		      mysql_read_default_group
		      mysql_socket
		      mysql_ssl
		      mysql_ssl_client_key
		      mysql_ssl_client_cert
		      mysql_ssl_ca_file
		      mysql_ssl_ca_path
		      mysql_ssl_cipher
		      mysql_local_infile);


sub connect_data {

  my $class = shift;

  my %p = @_;

  %p = validate( @_, { driver => { type => SCALAR },
		       dbname => { type => SCALAR },
		       user => { type => SCALAR | UNDEF,
				 optional => 1 },
		       password => { type => SCALAR | UNDEF,
				     optional => 1 },
		       host => { type => SCALAR | UNDEF,
				 optional => 1 },
		       port => { type => SCALAR | UNDEF,
				 optional => 1 },
		       map { $_ => 0 } (@optional_mysql_, @DBIx::DBH::attr)
		     } );

  my $dsn = "DBI:$p{driver}:$p{dbname}";
  foreach ( qw( host port ) ) {
    $dsn .= ";$_=$p{$_}" if (defined $p{$_});
  }


  foreach my $k ( grep { /^mysql/ } sort keys %p )
    {
      $dsn .= ";$k=$p{$k}";
    }


  ($dsn, $p{user}, $p{password}, $p{attr});

}


1;

=head1 NAME

  DBIx::DBH::mysql - mysql DBI database handle connection specifics.

=head1 connect()

L<DBIx::DBH> covered the options available to any database driver. Here we
list optional arguments supported by MySQL. Please read L<DBD::mysql> for 
details. All of these arguments apply to the DSN.

=over 4

=item * mysql_client_found_rows
=item * mysql_compression
=item * mysql_connect_timeout
=item * mysql_read_default_file
=item * mysql_read_default_group
=item * mysql_socket
=item * mysql_ssl
=item * mysql_ssl_client_key
=item * mysql_ssl_client_cert
-item * mysql_ssl_ca_file
-item * mysql_ssl_ca_path
-item * mysql_ssl_cipher
-item * mysql_local_infile


=back

=cut


lib/DBIx/DBH/SQLite.pm
package DBIx::DBH::SQLite;

use base qw(DBIx::DBH);
use Params::Validate qw( :all );

sub connect_data {
  my $class = shift;
  my %p = validate_with( 
            params => \@_,
            spec   => { 
                driver   => { type => SCALAR },
                dbname   => { type => SCALAR },
                attr     => { type => SCALAR | UNDEF, optional => 1 },
                user     => { type => SCALAR | UNDEF, optional => 1 },
                password => { type => SCALAR | UNDEF, optional => 1 },
                host     => { type => SCALAR | UNDEF, optional => 1 },
                port     => { type => SCALAR | UNDEF, optional => 1 },
		     },
             # allow_extra => 1,  
         );

  my $dsn = "DBI:$p{driver}:$p{dbname}";

  return ($dsn, $p{user}, $p{password}, $p{attr});

}


1;

=head1 NAME

  DBIx::DBH::SQLite - SQLite DBI database handle connection specifics.

=head1 connect()

dbname is probably the only one param you need. We pass through C<user>, C<password>,
and C<attr>, ignoring other attributes.

=head1 SEE ALSO

L<DBIx::DBH>
L<DBD::SQLite>

=head1 AUTHOR

Mark Stosberg, mark@summersault.com

=cut


lib/DBIx/DBH/Sybase.pm
package DBIx::DBH::Sybase;

use base qw(DBIx::DBH);
use Params::Validate qw( :all );

my @optional_ = qw(
syb_show_sql
syb_show_eed
syb_err_handler
syb_flush_finish
syb_dynamic_supported
syb_chained_txn
syb_quoted_identifier
syb_rowcount
syb_do_proc_status
syb_use_bin_0x
syb_binary_images
syb_oc_version
syb_server_version
syb_server_version_string
syb_failed_db_fatal
syb_no_child_con
syb_bind_empty_string_as_null
syb_cancel_request_on_error
syb_date_fmt
);

sub connect_data {

  my $class = shift;

  my %p = @_;

  %p = validate( @_, { driver => { type => SCALAR },
		       dbname => { type => SCALAR},
		       database => { type => SCALAR | UNDEF, 
				     optional => 1},
		       user => { type => SCALAR | UNDEF,
				 optional => 1 },
		       password => { type => SCALAR | UNDEF,
				     optional => 1 },
		       host => { type => SCALAR | UNDEF,
				 optional => 1 },
		       maxConnect => { type => SCALAR | UNDEF,
				 optional => 1 },
		       language => { type => SCALAR | UNDEF,
				       optional => 1 },
		       charset => { type => SCALAR | UNDEF,
				       optional => 1 },
		       packetSize => { type => SCALAR | UNDEF,
				       optional => 1 },
		       interfaces => { type => SCALAR | UNDEF,
				       optional => 1 },
		       timeout => { type => SCALAR | UNDEF,
				       optional => 1 },
		       loginTimeout => { type => SCALAR | UNDEF,
				    optional => 1 },
		       scriptName  => { type => SCALAR | UNDEF,
				    optional => 1 },
		       hostname  => { type => SCALAR | UNDEF,
				    optional => 1 },
		       tdsLevel  => { type => SCALAR | UNDEF,
				    optional => 1 },
		       encryptPassword   => { type => SCALAR | UNDEF,
				    optional => 1 },
		       kerberos   => { type => SCALAR | UNDEF,
				    optional => 1 },

		       sslCAFile  => { type => SCALAR | UNDEF,
				    optional => 1 },
		       bulkLogin  => { type => SCALAR | UNDEF,
				    optional => 1 },
		       port => { type => SCALAR | UNDEF,
				 optional => 1 },
		     
		       map { $_ => 0 } (@optional_, @DBIx::DBH::attr)
		     } );

  my $dsn = "dbi:$p{driver}:server=$p{host}";

  # optional DSN configuration options. 
  foreach ( qw(port database maxConnect language charset packetSize interfaces timeout loginTimeout scriptName hostname tdsLevel encryptPassword kerberos sslCAFile bulkLogin)) 
  {
      $dsn .= ";$_=$p{$_}" if (defined $p{$_});
  }
  
  # attributes that get passed in via an anonymous hash to the connect routine.
  foreach my $k ( grep { /^syb_/ } sort keys %p ) 
  {
      $p{attr}{$k} = $p{$k};
  }
  
  ($dsn, $p{user}, $p{password}, $p{attr});

}



1;

=head1 NAME

  DBIx::DBH::Sybase - Sybase DBI database handle connection specifics.

=head1 connect()

L<DBIx::DBH> covered the options available to any database driver as 
specified in L<DBI>. Here we
list optional arguments supported by Sybase. Please read L<DBD::Sybase> for
details

=head2 DSN options

=over 4

=item * maxConnect

=item * language

=item * charset

=item * packetSize

=item * interfaces

=item * timeout

=item * loginTimeout

=item * scriptName

=item * hostname

=item * tdsLevel

=item * encryptPassword

=item * kerberos

=item * sslCAFile

=item * bulkLogin


=back

=head2 Database handle attributes

=over 4


=item * syb_show_sql

=item * syb_show_eed

=item * syb_err_handler

=item * syb_flush_finish

=item * syb_dynamic_supported

=item * syb_chained_txn

=item * syb_quoted_identifier

=item * syb_rowcount

=item * syb_do_proc_status

=item * syb_use_bin_0x

=item * syb_binary_images

=item * syb_oc_version

=item * syb_server_version

=item * syb_failed_db_fatal

=item * syb_no_child_con

=item * syb_bind_empty_string_as_null

=item * syb_cancel_request_on_error

=item * syb_date_fmt

=back

=head1 SEE_ALSO

L<DBD::Sybase> to see what all of these options do.

=cut


blib/lib/DBIx/DBH/Pg.pm
package DBIx::DBH::Pg;

use base qw(DBIx::DBH);
use Params::Validate qw( :all );

my @optional_pg_ = qw(pg_enable_utf8 
		      pg_auto_escape 
		      pg_bool_tf);

sub connect_data {

  my $class = shift;

  my %p = @_;

  %p = validate( @_, { driver => { type => SCALAR },
		       dbname => { type => SCALAR },
		       user => { type => SCALAR | UNDEF,
				 optional => 1 },
		       password => { type => SCALAR | UNDEF,
				     optional => 1 },
		       host => { type => SCALAR | UNDEF,
				 optional => 1 },
		       port => { type => SCALAR | UNDEF,
				 optional => 1 },
		       options => { type => SCALAR | UNDEF,
				    optional => 1 },
		       tty => { type => SCALAR | UNDEF,
				optional => 1 },
		       map { $_ => 0 } (@optional_pg_, @DBIx::DBH::attr)
		     } );

  my $dsn = "dbi:$p{driver}:dbname=$p{dbname}";
  foreach ( qw( host port options tty ) ) {
    $dsn .= ";$_=$p{$_}" if (defined $p{$_});
  }

  foreach my $k ( grep { /^pg_/ } sort keys %p ) {
    $p{attr}{$k} = $p{$k};
  }

  ($dsn, $p{user}, $p{password}, $p{attr});

}



1;

=head1 NAME

  DBIx::DBH::Pg - Pg DBI database handle connection specifics.

=head1 connect()

L<DBIx::DBH> covered the options available to any database driver as 
specified in L<DBI>. Here we
list optional arguments supported by Pg. Please read L<DBD::Pg> for
details

=head2 DSN options

=over 4

=item * options

=item * tty

=back

=head2 Database handle attributes

=over 4

=item * pg_auto_escape

=item * pg_enable_utf8

=item * pg_bool_tf

=back

=cut


lib/DBIx/DBH/Legacy.pm
package DBIx::DBH::Legacy;

use strict;
use warnings;



use DBI;
use Params::Validate qw( :all );

our $VERSION = '0.2';

our @attr = qw
  (  
   dbi_connect_method
   Warn

   _Active
   _Executed
   _Kids
   _ActiveKids
   _CachedKids
   _CompatMode

   InactiveDestroy
   PrintWarn
   PrintError
   RaiseError
   HandleError
   HandleSetErr

   _ErrCount

   ShowErrorStatement
   TraceLevel
   FetchHashKeyName
   ChopBlanks
   LongReadLen
   LongTruncOk
   TaintIn
   TaintOut
   Taint
   Profile
   _should-add-support-for-private_your_module_name_*


   AutoCommit

   _Driver
   _Name
   _Statement

   RowCacheSize

   _Username
  );

# Preloaded methods go here.

Params::Validate::validation_options(allow_extra => 1);

sub connect {

  my @connect_data = connect_data(@_);

  my $dbh;
  eval
    {
      $dbh = DBI->connect( @connect_data );
    };

  die $@ if $@;
  die 'Unable to connect to database' unless $dbh;

  return $dbh;

}

sub dbi_attr {
  my ($h, %p) = @_;

  $h = {} unless defined $h;

  for my $attr (@attr) {
    if (exists $p{$attr}) {
#      warn "$attr = $p{$attr};";
      $h->{$attr} = $p{$attr};
    }
  }

  $h;
}

sub connect_data {

  my $class = shift;
  my %p = @_;

  my $subclass = "DBIx::DBH::$p{driver}";
  eval "require $subclass";
  die "unable to require $subclass to support the $p{driver} driver.\n" if $@;

  my ($dsn, $user, $pass, $attr) = $subclass->connect_data(@_);
  $attr = dbi_attr($attr, %p);

  ($dsn, $user, $pass, $attr)

}

sub form_dsn {

  (connect_data(@_))[0];

}


1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

 DBIx::DBH - helper for DBI connection data (form dsn, etc)

=head1 SYNOPSIS

 use DBIx::DBH;

 my %opt = (tty => 1) ;
 my %dat = ( 
     driver => 'Pg',
     dbname => 'db_terry',
     user => 'terry',
     password => 'markso'
 );

 my $dbh = DBIx::DBH->connect(%dat, %opt) ; # yes, two hashes, not hrefs!

=head1 ABSTRACT

L<DBIx::DBH> is designed to facilitate and validate the process of creating 
L<DBI> database connections.
It's chief and unique contribution to this set of modules on CPAN is that
it forms the DSN string for you, regardless of database driver. Another thing 
about this module is that
it takes a flat Perl hash 
as input, making it ideal for converting HTTP form data 
and or config file information into DBI database handles. It also can form
DSN strings for both major free databases and is subclassed to support
extension for other databases.

DBIx::DBH provides rigorous validation on the input parameters via
L<Params::Validate>. It does not
allow parameters which are not defined by the DBI or the database driver
driver into the hash.

It provides support for MySQL, Postgres and Sybase (thanks to Rachel Richard
for the Sybase support).

=head1 Motivation

This module does not appear to be very useful at first. But it has it's place.
Let's see why.

=head2 Simple, robust DSN formation

=head3 Simple

Let's take a look at a L<DBI> connection string:

  DBI->connect("dbi:mysql:database=sakila;host=localhost;post=3306",
       $username, $password);

Now, notice: how the C<dsn> contains a lot of subelements:

=over 4

=item   1. dbi

=item   2. mysql

=item   3. database

=item   4. host

=item   5. port 

=cut

With this module, you simply specify those sub-elements in a hash:

  my %dat = ( 
     driver => 'mysql',
     dbname => 'sakila',
     user => 'username',
     password => 'pass'
  );

This is much more high-level.

So, the first win is that you get to be DWIM instead of DWIS.


=head3 Robust

This module is robust. It uses L<Params::Validate> to make sure that
what you supply is valid.

=head2 Easier interaction with APIs

=head3 Rose::DB::register_db() expects sub-components of a DSN

If you take a look at a call to C<register_db>:

L<http://search.cpan.org/~jsiracusa/Rose-DB-0.758/lib/Rose/DB/Tutorial.pod#Just_one_data_source>

you will notice that it requires the sub-components of the DSN. So, 
ideally you would be able to keep your connection data as a set of
sub-components and supply it to L<Rose::DB> but when you want to connect
directly to L<DBI>, you could do that also.

This module is the solution for this dilemma as well.

=head3 Alzabo and DBIx::AnyDBD have alternative connection syntaxes

Alternative connection syntaxes such as L<DBIx::AnyDBD> or 
L<Alzabo> can make use of the C<connect_data> API call


=head1 API

=head2 $dbh = connect(%params)

C<%params> requires the following as keys:

=over 4

=item * driver : the value matches /\a(mysql|Pg)\Z/ (case-sensitive).

=item * dbname : the value is the name of the database to connect to

=back

C<%params> can have the following optional parameters

=over 4

=item * user

=item * password

=item * host

=item * port

=back

C<%params> can also have parameters specific to a particular database
driver. See
L<DBIx::DBH::Sybase>,
L<DBIx::DBH::mysql> and L<DBIx::DBH::Pg> for additional parameters
acceptable based on database driver.

=head2 ($dsn, $user, $pass, $attr) = connect_data(%params)

C<connect_data> takes the same arguments as C<connect()> but returns
a list of the 4 arguments required by the L<DBI> C<connect()>
function. This is useful for working with modules that have an
alternative connection syntax such as L<DBIx::AnyDBD> or 
L<Alzabo>.

=head2 $dsn = form_dsn(%params)

C<form_dsn> takes the same arguments as C<connect()> but returns
only the properly formatted DSN string. This is also 
useful for working with modules that have an
alternative connection syntax such as L<DBIx::AnyDBD> or 
L<Alzabo>.

=head1 ADDING A DRIVER

Simply add a new driver with a name of C<DBIx::DBH::$Driver>, where
C<$Driver> is a valid DBI driver name.

=back

=head1 SEE ALSO

=over

=item * L<Config::DBI>

=item * L<DBIx::Connect>

=item * L<DBIx::Password>

=item * L<Ima::DBI>

=back

=head2 Links

=head3 "Avoiding compound data in software and system design"

L<http://perlmonks.org/?node_id=835894>


=head1 TODO

=over

=item * use a singleton object

The current API for DBIx::DBH requires passing in the connection data 
hash to each API function. The data hash should be bound to a singleton
object and all methods should resource it.

A good set of L<Moose> roles inspired by L<MooseX::Role::DBIx::Connector>
or L<DBIx::Roles> might be in order.

=item * use DBIx::Connector

L<DBIx::Connector> is an excellent module for reusing DBI database connections.
This module should optionally connect to DBI via that instead of directly.

=item * expose parm validation info:

 > 
 > It would be nice if the parameter validation info was exposed in some 
 > way, so that an interactive piece of software can ask a user which 
 > driver they want, then query your module for a list of supported 
 > parameters, then ask the user to fill them in. (Perhaps move the hash 
 > of validation parameters to a new method named valid_params, and then 
 > have connect_data call that method and pass the return value to 
 > validate?)

=cut

=head1 AUTHOR

Terrence Brannon, E<lt>bauhaus@metaperl.comE<gt>

Sybase support contributed by Rachel Richard.

Mark Stosberg did all of the following:

=over

=item * contributed Sqlite support

=item * fixed a documentation bug

=item * made DBIx::DBH more scaleable

Says Mark: "Just as DBI needs no modifications for a new driver to work,
neither should this module.

I've attached a patch which refactors the code to address this.

Rather than relying on a hardcoded list, it tries to 'require' the
driver, or dies with a related error message.

This could lower your own maintenance effort, as others can publish
additional drivers directly without requiring a new release of
DBIx::DBH for it to work."

L<http://rt.cpan.org/Ticket/Display.html?id=18026>

=back



Substantial suggestions by M. Simon Ryan Cavaletto.

=head1 SOURCECODE

L<http://github.com/metaperl/dbix-dbh>

=head1 COPYRIGHT AND LICENSE

Copyright (C) by Terrence Brannon

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.


=cut
